bit_manipulation/hamming_code_generator.py

```diff
@@ -0,0 +1,114 @@
+""" "
+Author: Ayush Chakraborty
+Date: 6th October 2024
+
+generate (15,11) hamming encoded bits gives 11 bit data
+
+input: 11 bit binary number with type string
+return: 16 bit binary hamming encoded number returned in string form
+"""
+
+from functools import reduce
+
+
+def hamming_15_11(number: str) -> str:
+    """
+    Performs parity checks to assign values to the redundant bits,
+    in the 16 bit number
+    returned, bits at index 0, 1, 2, 4, 8 are redundant bits used for checking
+
+    Hamming generated 16 bits generated from the 11 bit binary number can only detect
+    and change a single bit change, but can only detect more than a
+    single bit change
+
+    for more theoretical knowledege about Hamming codes, refer:
+    https://www.youtube.com/watch?v=X8jsijhllIA&t=143s
+    by 3B1B YT channel
+
+    >>> hamming_15_11("00110001110")
+    '0010101110001110'
+    >>> hamming_15_11("10110000110")
+    '0101001100000110'
+    >>> hamming_15_11("10111100110")
+    '0011001101100110'
+    >>> hamming_15_11("10001000010")
+    '0001100001000010'
+
+    """
+    if len(number) == 11:
+        digits = [0 if number[i] == "0" else 1 for i in range(len(number))]
+        total_num_1 = sum(digits)
+        hamming_digits = [0 for i in range(16)]
```

Comment:
Cool approach. I feel like the reduce + lambda can be hard to follow along at times. I believe the following has the same result and is more intuitive to follow:

```
parity_positions = {
    1: [0, 1, 3, 4, 6, 8, 10],
    2: [0, 2, 3, 5, 6, 9, 10],
    4: [1, 2, 3, 7, 8, 9, 10],
    8: [4, 5, 6, 7, 8, 9, 10],
}

parity_bits = {}
for parity_pos, data_indices in parity_positions.items():
    parity = 0
    for idx in data_indices:
        parity ^= digits[idx]
    parity_bits[parity_pos] = parity
```

Reply:
Thanks! Introducing similar changes